File: project_euler_code/javascript/37.js
function isPrime(n) {
  if (n <= 1)
    return false;
  var s = Math.floor(Math.sqrt(n)) + 1;
  for (var i = 2; i < s; i++)
    if (n % i === 0)
      return false;
  return true;
}

function isLeftTruncatable(n) {
  var r = 0, p = 1;
  while (n > 0) {
    r += p * (n % 10);
    p *= 10;
    if (!isPrime(r))
      return false;
    n = Math.floor(n / 10);
  }
  return true;
}

function pe37() {
  var trunc = [];
  function recurse(left) {
    left *= 10;
    [1, 3, 7, 9].forEach(function(n) {
      var p = left + n;
      if (isPrime(p)) {
        recurse(p);
        if ((n === 3 || n === 7) && isLeftTruncatable(p))
          trunc.push(p);
      }
    });
  }
  [2, 3, 5, 7].forEach(recurse);
  return trunc.reduce(function(a, b) { return a + b; });
}

var startTime = Date.now();
console.log(pe37());
console.log('finished in ' + (Date.now() - startTime) / 1000 + 's');

File: project_euler_code/c/37.c
#include <stdio.h>
#include <stdbool.h>
#include <math.h>

static bool is_prime(int n) {
    if (n <= 2)
        return n == 2;
    int s = sqrt(n) + 1;
    for (int i = 3; i < s; i += 2)
        if (n % i == 0)
            return false;
    return true;
}

static bool is_left_truncatable(int n) {
    int r = 0, p = 1;
    while (n > 0) {
        r += p * (n % 10);
        if (!is_prime(r))
            return false;
        p *= 10;
        n /= 10;
    }
    return true;
}

static inline int recurse(int left) {
    left *= 10;
    int s = 0;
    for (int i = 0; i < 4; i++) {
        int n = 3 * i + ((i & 1) ^ 1);
        int p = left + n;
        if (is_prime(p)) {
            s += recurse(p);
            if ((n == 3 || n == 7) && is_left_truncatable(p))
                s += p;
        }
    }
    return s;
}

static int pe37() {
    return recurse(2) + recurse(3) + recurse(5) + recurse(7);
}

int main(void) {
    printf("%d\n", pe37());
}
